### ISSUE TITLE
Host suddenly stops receiving new messages without recovering

### ISSUE THREAD

**Issue by royyeah** [Azure/azure-event-hubs-java#377]
## Actual Behavior
I am using an EventProcessorHost with basic setup as explained [here](https://github.com/Azure/azure-event-hubs-java/blob/1.0.0/azure-eventhubs-eph/Overview.md). Recently we ran into problems where the host stops receiving new messages because it did not recover correctly. When we restarted the application it continued processing again.

## Expected Behavior
I think the warnings below are the result of event hub partitions being moved across nodes. I would expect the event processor host to automatically recover from this.

Is this fixed in azure-eventhubs 1.1.0? I see this contains bugfixes [#372](https://github.com/Azure/azure-event-hubs-java/pull/372) and [#373](https://github.com/Azure/azure-event-hubs-java/pull/373) which sound like potential fixes.

## Versions  
- OS platform and version: Debian 8
- Maven package version or commit ID:
  - azure-eventhubs: 1.0.2
  - azure-eventhubs-eph: 2.0.1

## More details
Log excerpt:

```
2018-09-02 08:32:45,500 WARN  [pool-2-thread-9] c.m.a.e.i.ReceivePump - Receive pump for partition (8) exiting after receive exception java.util.concurrent.CompletionException: com.microsoft.azure.eventhubs.EventHubException: No connection handler was found for virtual host '19216'. Remote container id is '7A2E0A0515A9427F82D1ACF7F25A024D_G1'. TrackingId:cd407ddfef2847d0a54fafa43b6c42bc_B13, SystemTracker:SharedConnectionListener, Timestamp:9/2/2018 8:32:45 AM TrackingId:2a1dd72a1af240b295ecaa259ca8928a_G1, SystemTracker:gateway5, Timestamp:9/2/2018 8:32:45 AM, errorContext[NS: enxnsdatalakealphaprd.servicebus.windows.net, PATH: ehsalvador7readings/ConsumerGroups/cgcassandra/Partitions/8, REFERENCE_ID: cf942a_928a_G1_1535877165274, PREFETCH_COUNT: 300, LINK_CREDIT: 0, PREFETCH_Q_LEN: 0]
2018-09-02 08:32:45,500 WARN  [pool-2-thread-9] c.m.a.e.PartitionPump - host salvador-to-cassandra-758ed9cf-722d-4674-b58d-a68803f15738: 8: EventHub client error: java.util.concurrent.CompletionException: com.microsoft.azure.eventhubs.EventHubException: No connection handler was found for virtual host '19216'. Remote container id is '7A2E0A0515A9427F82D1ACF7F25A024D_G1'. TrackingId:cd407ddfef2847d0a54fafa43b6c42bc_B13, SystemTracker:SharedConnectionListener, Timestamp:9/2/2018 8:32:45 AM TrackingId:2a1dd72a1af240b295ecaa259ca8928a_G1, SystemTracker:gateway5, Timestamp:9/2/2018 8:32:45 AM, errorContext[NS: enxnsdatalakealphaprd.servicebus.windows.net, PATH: ehsalvador7readings/ConsumerGroups/cgcassandra/Partitions/8, REFERENCE_ID: cf942a_928a_G1_1535877165274, PREFETCH_COUNT: 300, LINK_CREDIT: 0, PREFETCH_Q_LEN: 0]
2018-09-02 08:32:45,500 WARN  [pool-2-thread-9] c.m.a.e.PartitionPump - host salvador-to-cassandra-758ed9cf-722d-4674-b58d-a68803f15738: 8: EventHub client error continued
java.util.concurrent.CompletionException: com.microsoft.azure.eventhubs.EventHubException: No connection handler was found for virtual host '19216'. Remote container id is '7A2E0A0515A9427F82D1ACF7F25A024D_G1'. TrackingId:cd407ddfef2847d0a54fafa43b6c42bc_B13, SystemTracker:SharedConnectionListener, Timestamp:9/2/2018 8:32:45 AM TrackingId:2a1dd72a1af240b295ecaa259ca8928a_G1, SystemTracker:gateway5, Timestamp:9/2/2018 8:32:45 AM, errorContext[NS: enxnsdatalakealphaprd.servicebus.windows.net, PATH: ehsalvador7readings/ConsumerGroups/cgcassandra/Partitions/8, REFERENCE_ID: cf942a_928a_G1_1535877165274, PREFETCH_COUNT: 300, LINK_CREDIT: 0, PREFETCH_Q_LEN: 0]
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:593)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
	at com.microsoft.azure.eventhubs.impl.ExceptionUtil.completeExceptionally(ExceptionUtil.java:86)
	at com.microsoft.azure.eventhubs.impl.MessageReceiver.drainPendingReceives(MessageReceiver.java:404)
	at com.microsoft.azure.eventhubs.impl.MessageReceiver.onError(MessageReceiver.java:390)
	at com.microsoft.azure.eventhubs.impl.MessageReceiver.onClose(MessageReceiver.java:621)
	at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.processOnClose(BaseLinkHandler.java:50)
	at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.handleRemoteLinkClosed(BaseLinkHandler.java:70)
	at com.microsoft.azure.eventhubs.impl.BaseLinkHandler.onLinkRemoteClose(BaseLinkHandler.java:36)
	at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:176)
	at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
	at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
	at com.microsoft.azure.eventhubs.impl.MessagingFactory$RunReactor.run(MessagingFactory.java:483)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.microsoft.azure.eventhubs.EventHubException: No connection handler was found for virtual host '19216'. Remote container id is '7A2E0A0515A9427F82D1ACF7F25A024D_G1'. TrackingId:cd407ddfef2847d0a54fafa43b6c42bc_B13, SystemTracker:SharedConnectionListener, Timestamp:9/2/2018 8:32:45 AM TrackingId:2a1dd72a1af240b295ecaa259ca8928a_G1, SystemTracker:gateway5, Timestamp:9/2/2018 8:32:45 AM, errorContext[NS: enxnsdatalakealphaprd.servicebus.windows.net, PATH: ehsalvador7readings/ConsumerGroups/cgcassandra/Partitions/8, REFERENCE_ID: cf942a_928a_G1_1535877165274, PREFETCH_COUNT: 300, LINK_CREDIT: 0, PREFETCH_Q_LEN: 0]
	at com.microsoft.azure.eventhubs.impl.ExceptionUtil.toException(ExceptionUtil.java:58)
	at com.microsoft.azure.eventhubs.impl.MessageReceiver.onClose(MessageReceiver.java:620)
	... 15 more
2018-09-02 08:32:45,503 INFO  [pool-2-thread-5] c.m.a.e.i.ReceivePump - Stopping receive pump for partition (8) as pump ran into errors.
```

**Comment by SreeramGarlapati** [comment]
That's right #373 - fixes this issue. May I request you to try latest version of EPH with latest version (`1.1.0`) of JavaClient and **ack** here that they are compatible and helps with this issue..

**Comment by royyeah** [comment]
I have changed to `1.1.0`, however since I have no idea when partitions are being moved across nodes at Azure, which makes it hard to verify if it solved the issue. Do you have an idea how we can verify this?

**Comment by SreeramGarlapati** [comment]
@royyeah - if you enable warn level logs for EventHub client package (along with EPH) - you would see these errors in the EventHub client logs `No connection handler was found for virtual host '19216'. Remote container id is '7A2E0A0515A9427F82D1ACF7F25A024D_G1'.` and if you dont see corresponding partition move warn logs in `EPH` logs - there it is!

**Comment by royyeah** [comment]
@SreeramGarlapati - thanks, I will keep an eye on it.

**Comment by SreeramGarlapati** [comment]
Thanks a lot for `ack`ing that these two packages go well with each other. I was able to recommend to couple other customers based on that!
Am closing this as the code fix for this issue is in. Please feel free to reopen - if you see this again - @royyeah .

**Comment by SreeramGarlapati** [comment]
@royyeah - thanks a lot for reporting the issue and presenting the logs. 
Earlier, I got carried away by the error message & parts of the stack trace & didnot realize a nuance - in the error stack trace (I guess I was under the influence of `Confirmation bias` :)). 
#373 - implements retry mechanism on `createReceiver` call - whereas, the issue reported here is on the `PartitionReceiver.Receive` API call. It would not have fixed this - due to similarities in the error message and parts of stacktrace - I did not realize that - appologies about that.
The actual issue here is that: `PartitionReceiver.Receive` API - should always return `null` - when there are `Transient` errors (i.e., when EventHubException.isTransient()` is true) - and should never throw transient exceptions to the end user. However, due to a bug - in some cases - this is not happenning. And as a result - it throws this exception on the pump. Pump doesn't expect that and quits.
Since, EventHubs partitions are generally highly available and rarely have the downtime of over a minute - this was hard to reproduce. This is fixed in #381 (included as part of 1.2.0 release).
https://github.com/Azure/azure-event-hubs-java/blame/1.2.0/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageReceiver.java#L433
Please go ahead and upgrade and enable warn level logs and open a new issue like you did earlier - if you see it again...

**Comment by royyeah** [comment]
@SreeramGarlapati - great, thanks, I will try the new version and let you know if I see it again.
